### PR TITLE
Fix Constraint Planning Segfault

### PR DESCRIPTION
**Pull request by MarqRazz**
### Description
This PR is an attempt to understand and hopefully fix a segfault I have been running into when asking MoveIt to solve constrained motion as reported in #2103.

While debugging we have found that the `ompl::base::PlannerPtr` generated on the first planning request is going out of scope causing the next planning request to fault the program. As you can see in my initial commit I have removed the ability for the planning allocator to reuse the planner between requests which fixes the segfault. This is not ideal if the user is attempting to build a [roadmap database](https://moveit.picknik.ai/humble/doc/examples/ompl_interface/ompl_interface_tutorial.html#persistent-roadmaps) because it would only contain the map of a single plan if written to disk, or thrown away on the next request. The planner is created [on this line](https://github.com/MarqRazz/moveit2/blob/b19e25adc724b27ffef350313bd8f666946916c8/moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp#L190) by passing it the `SpaceInformationPtr` that is created [here](https://github.com/MarqRazz/moveit2/blob/b19e25adc724b27ffef350313bd8f666946916c8/moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp#L375-L378). If I'm understanding OMPL correctly this `SpaceInformationPtr` is being moved to the Planner object [here](https://github.com/ompl/ompl/blob/main/src/ompl/base/src/Planner.cpp#L44-L49) and the `ConfiguredPlannerAllocator` containing it should be kept alive/in context by the [known_planners_](https://github.com/ros-planning/moveit2/blob/acbf3770763e58e4f86ce0dea251a50cd4987fe6/moveit_planners/ompl/ompl_interface/include/moveit/ompl_interface/planning_context_manager.h#L232) object of the Planning Context Manager.

I would really appreciate any help in identifying why the pointer, or objects inside of it, is going out of scope of if my understanding of the program is incorrect! As I have been learning this section of the software I have also wondered if the current implementation is worth fixing / there is a better way to preserve the planning scene between motion requests. Because the planner is constructed with the initial planning request goal's [path_constraint](https://github.com/MarqRazz/moveit2/blob/b19e25adc724b27ffef350313bd8f666946916c8/moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp#L366-L367) does this mean that if it is reused it could have invalid constraints specified if they change between requests?

### CODE REVIEW

**Comment by henningkayser** [comment]
Please see my comment here https://github.com/ros-planning/moveit2/issues/2103#issuecomment-1523720752

**Comment by mamoll** [comment]
If this fixes the crash, then this seems like a useful workaround that should be merged. Recreating a planner instance from planner data adds a bit of overhead, but shouldn't be too bad. I'd add a comment along these lines so that we don't forget why this is done:
```cpp
// FIXME: make reusing PlannerPtr not crash, so that we don't have to reconstruct a PlannerPtr instance
```

File: moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp [hunk 1]
```diff
@@ -117,11 +118,17 @@ ompl::base::PlannerPtr MultiQueryPlannerAllocator::allocatePlanner(const ob::Spa
...
   {
-    // If we already have an instance, use that one
+    // If we already have an instance, reuse it's planning data
+    // FIXME: make reusing PlannerPtr not crash, so that we don't have to reconstruct a PlannerPtr instance
     auto planner_map_it = planners_.find(new_name);
     if (planner_map_it != planners_.end())
     {
-      return planner_map_it->second;
+      ob::PlannerData data(si);
+      planner_map_it->second->getPlannerData(data);
```
Comment (henningkayser) [inline moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp:127]: I'm confused now. How can we access the planner data if the planner instance is invalid? Or are there actually any members inside the planner instance that go out of scope? This version looks like it should produce the same segfaults

**Review by henningkayser — APPROVED** [review]
I'm fine with merging this as a temporary solution. The overhead of creating a new planner instance is really not too much of an issue, but we should keep on looking into the underlying cause anyway.

**Comment by MarqRazz** [comment]
Just so everyone knows, I have done some profiling and re-constructing the planner is taking a little under 0.2 seconds which is not ideal for application that are going for fast cycle times!
```
ompl::geometric::LazyPRM::LazyPRM(const base::PlannerData &data, bool starStrategy)
*** Profiling statistics. Total counted time : 0.181746 seconds
Blocks of time:
lazyPRMConstructor: 0.181721s (99.9865%), [0.181721s --> 0.181721 s], 1 parts, 0.181721 s on average
Unaccounted time : 2.4583e-05 (0.013526 %)
```

I have also found that re-using the planning graph does not speed up LazyPRM. If you give it the exact same problem to solve over and over again the graph will increasing in size with each attempt but the planning time always uses the max allowed set by the user. This could be an error with my configuration so I will update this PR with more details if I can get it working better.

File: moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp [hunk 2]
```diff
@@ -117,11 +118,17 @@ ompl::base::PlannerPtr MultiQueryPlannerAllocator::allocatePlanner(const ob::Spa
   }
   if (multi_query_planning_enabled)
   {
-    // If we already have an instance, use that one
+    // If we already have an instance, reuse it's planning data
+    // FIXME: make reusing PlannerPtr not crash, so that we don't have to reconstruct a PlannerPtr instance
```
Comment (sjahr) [inline moveit_planners/ompl/ompl_interface/src/planning_context_manager.cpp:122]: Can you open a follow-up issue for this? With a summery of you findings and (if possible) a path forward so somebody else can take this up in the future?

**Review by sjahr — APPROVED** [review]
Let's merge this fix and open a follow-up issue :+1:

**Review by sjahr — APPROVED** [review]
Let's merge this fix and open a follow-up issue :+1:

**Comment by MarqRazz** [comment]
Can we back port this to Humble? I don't have permissions to ask mergify.

**Pull request merged**